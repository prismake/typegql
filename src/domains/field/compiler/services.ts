import {
  isOutputType,
  GraphQLType,
  GraphQLOutputType,
  GraphQLNonNull,
} from 'graphql';
import { FieldError } from '../index';

import { resolveTypeOrThrow, inferTypeOrThrow } from './fieldType';
import {
  mutationFieldsRegistry,
  queryFieldsRegistry,
  isSchemaRoot,
} from '~/domains/schema';

export function resolveRegisteredOrInferedType(
  target: Function,
  fieldName: string,
  forcedType?: any,
) {
  if (forcedType) {
    return resolveTypeOrThrow(forcedType, target, fieldName);
  }
  return inferTypeOrThrow(target, fieldName);
}

export function validateResolvedType(
  target: Function,
  fieldName: string,
  type: GraphQLType,
): type is GraphQLOutputType {
  if (!isOutputType(type)) {
    throw new FieldError(
      target,
      fieldName,
      `Validation of type failed. Resolved type must be a GraphQLOutputType.`,
    );
  }
  return true;
}

export function enhanceType(
  originalType: GraphQLOutputType,
  isNullable: boolean,
) {
  let finalType = originalType;
  if (!isNullable) {
    finalType = new GraphQLNonNull(finalType);
  }
  return finalType;
}

export function isRootFieldOnNonRootBase(base: Function, fieldName: string) {
  const isRoot = isSchemaRoot(base);
  if (isRoot) {
    return false;
  }
  if (mutationFieldsRegistry.has(base, fieldName)) {
    return true;
  }
  if (queryFieldsRegistry.has(base, fieldName)) {
    return true;
  }
  return false;
}
