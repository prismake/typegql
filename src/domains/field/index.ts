import { fieldsRegistry, FieldInnerConfig } from './registry';

export {
  FieldInnerConfig,
  fieldsRegistry,
  queryFieldsRegistry,
} from './registry';
export { compileAllFields, compileFieldConfig } from './compiler';
export { isQueryField } from './services';
export { FieldError } from './error';

export interface FieldOptions {
  description?: string;
  type?: any;
  name?: string;
  isNullable?: boolean;
}

export function Field(options?: FieldOptions): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: true,
      ...options,
    };

    const descriptor = Object.getOwnPropertyDescriptor(
      targetInstance,
      fieldName,
    );

    fieldsRegistry.set(targetInstance.constructor, fieldName, {
      isGetter: descriptor && !!descriptor.get,
      ...finalConfig,
    });
  };
}
