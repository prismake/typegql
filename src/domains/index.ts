export {
  ObjectType,
  compileObjectType,
  objectTypeRegistry,
} from './objectType';
export {
  InterfaceType,
  compileInterfaceType,
  interfaceTypeRegistry,
} from './interfaceType';
export {
  InputObjectType,
  compileInputObjectType,
  inputObjectTypeRegistry,
} from './inputObjectType';
export { Field } from './field';
export { InputField } from './inputField';
export { Arg } from './arg';
export { Inject, Context, Source, Info } from './inject';
export { registerEnum, enumsRegistry } from './enum';
export { Union, unionRegistry } from './union';
export { Before, After } from './hooks';
export {
  SchemaRoot,
  Schema,
  schemaRootsRegistry,
  compileSchema,
  Query,
  Mutation,
} from './schema';
