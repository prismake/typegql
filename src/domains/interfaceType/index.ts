import { compileInterfaceTypeWithConfig } from './compiler';
import { interfaceTypeRegistry } from './registry';

export { compileInterfaceType } from './compiler';
export { InterfaceTypeError } from './error';
export { interfaceTypeRegistry } from './registry';

export interface InterfaceTypeOptions {
  name?: string;
  description?: string;
}

export function InterfaceType(options?: InterfaceTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileInterfaceTypeWithConfig(target, config);
    interfaceTypeRegistry.set(target, outputTypeCompiler);
  };
}
