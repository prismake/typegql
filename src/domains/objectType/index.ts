import { compileObjectTypeWithConfig } from './compiler';
import { objectTypeRegistry } from './registry';

export { compileObjectType } from './compiler';
export { ObjectTypeError } from './error';
export { objectTypeRegistry } from './registry';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function ObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const config = { name: target.name, ...options };
    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, config);
    objectTypeRegistry.set(target, outputTypeCompiler);
  };
}
