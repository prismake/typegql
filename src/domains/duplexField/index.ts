import { fieldsRegistry, FieldInnerConfig } from '../field/registry';
import {
  inputFieldsRegistry,
  FieldInputInnerConfig,
} from '../inputField/registry';

export { FieldError } from './error';

export function DuplexField(
  options?: FieldInnerConfig &
    FieldInputInnerConfig & {
      inputNullable?: boolean;
    },
): PropertyDecorator {
  return (targetInstance: Object, fieldName: string) => {
    let isNullable = true;
    let inputNullable = true;
    if (options) {
      if (options.isNullable !== undefined) {
        isNullable = options.isNullable;
      }
      if (options.inputNullable !== undefined) {
        inputNullable = options.inputNullable;
      }
      delete options.isNullable;
      delete options.inputNullable;
    }

    const finalInputConfig: FieldInputInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable: inputNullable,
      ...options,
    } as any;

    const finalConfig: FieldInnerConfig = {
      property: fieldName,
      name: fieldName,
      isNullable,
      ...options,
    } as any;
    fieldsRegistry.set(targetInstance.constructor, fieldName, finalConfig);

    inputFieldsRegistry.set(
      targetInstance.constructor,
      fieldName,
      finalInputConfig,
    );
  };
}
