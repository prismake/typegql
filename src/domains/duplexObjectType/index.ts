import { objectTypeRegistry } from '../objectType/registry';
import { compileObjectTypeWithConfig } from '../objectType/compiler';
import { compileInputObjectTypeWithConfig } from '../inputObjectType/compiler';
import { inputObjectTypeRegistry } from '../inputObjectType/registry';
export { ObjectTypeError } from './error';

export interface ObjectTypeOptions {
  name?: string;
  description?: string;
}

export function DuplexObjectType(options?: ObjectTypeOptions): ClassDecorator {
  return (target: Function) => {
    const inputConfig = { name: target.name + 'Input', ...options };
    const outputConfig = { name: target.name, ...options };

    const inputTypeCompiler = () =>
      compileInputObjectTypeWithConfig(target, inputConfig);

    inputObjectTypeRegistry.set(target, inputTypeCompiler);

    const outputTypeCompiler = () =>
      compileObjectTypeWithConfig(target, outputConfig);
    objectTypeRegistry.set(target, outputTypeCompiler);
  };
}
