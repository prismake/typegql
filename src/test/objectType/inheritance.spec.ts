import { GraphQLString, GraphQLNonNull, printType } from 'graphql';
import { InterfaceType, ObjectType, compileObjectType, Field } from '~/domains';
import { getClassWithAllParentClasses } from '~/services/utils';

describe('Type inheritance', () => {
  it('Will pass fields from parent class', () => {
    class Base {
      @Field() baseField: string;
    }

    @ObjectType()
    class Foo extends Base {}

    const { baseField } = compileObjectType(Foo).getFields();

    expect(baseField.type).toEqual(GraphQLString);
  });

  it('Will overwrite fields in child class', () => {
    class Base {
      @Field() foo: string;
      @Field() bar: string;
    }

    @ObjectType()
    class Foo extends Base {
      @Field({ isNullable: false })
      foo: string;
    }

    const { foo, bar } = compileObjectType(Foo).getFields();

    expect(bar.type).toEqual(GraphQLString);
    expect(foo.type).toEqual(new GraphQLNonNull(GraphQLString));
  });

  it('picks up all the properties even for long chain of extended classes', async () => {
    @ObjectType()
    class Vehicle {
      @Field() passengers: string;
    }

    @ObjectType()
    class Car extends Vehicle {
      @Field() doorCount: number;
    }

    @ObjectType()
    class Lamborghini extends Car {
      @Field() speed: string;
    }
    const compiled = compileObjectType(Lamborghini);

    const fields = compiled.getFields();

    expect(fields).toHaveProperty('passengers');
    expect(fields).toHaveProperty('doorCount');
    expect(fields).toHaveProperty('speed');
    expect(getClassWithAllParentClasses(Lamborghini)).toHaveLength(3);
  });

  it('implements an interface when extending by an InterfaceType', async () => {
    @InterfaceType()
    class Vehicle {
      @Field() passengers: string;
    }

    @ObjectType()
    class Car extends Vehicle {
      @Field() doorCount: number;
    }

    const compiled = compileObjectType(Car);

    const fields = compiled.getFields();

    expect(fields).toHaveProperty('passengers');
    expect(fields).toHaveProperty('doorCount');
    expect(getClassWithAllParentClasses(Car)).toHaveLength(2);

    expect(compiled.getInterfaces()).toHaveLength(1);
    expect(printType(compiled)).toMatchSnapshot();
  });

  it('can use multiple interfaces from InterfaceType ancestors', async () => {
    @InterfaceType()
    class Entity {
      @Field() name: string;
    }

    @InterfaceType()
    class Vehicle extends Entity {
      @Field() passengers: string;
    }

    @ObjectType()
    class Car extends Vehicle {
      @Field() doorCount: number;
    }

    const compiled = compileObjectType(Car);

    const fields = compiled.getFields();
    expect(fields).toHaveProperty('name');
    expect(fields).toHaveProperty('passengers');
    expect(fields).toHaveProperty('doorCount');
    expect(getClassWithAllParentClasses(Car)).toHaveLength(3);

    expect(compiled.getInterfaces()).toHaveLength(2);
    expect(printType(compiled)).toMatchSnapshot();
  });
});
