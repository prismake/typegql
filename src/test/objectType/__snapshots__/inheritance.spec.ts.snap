// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Type inheritance can use multiple interfaces from InterfaceType ancestors 1`] = `
"type Car implements Entity & Vehicle {
  name: String
  passengers: String
  doorCount: Float
}"
`;

exports[`Type inheritance implements an interface when extending by an InterfaceType 1`] = `
"type Car implements Vehicle {
  passengers: String
  doorCount: Float
}"
`;
